### DISCUSSION TITLE
Latency

### DISCUSSION THREAD

**Discussion by midimusic** [GrandOrgue/grandorgue#1651] [General]
The subject is Latency.

What is the scientific definition of latency with regard to organs? And how is it to be properly measured? If latency cannot be measured then latency figures quoted for organs are almost meaningless.

In simplistic terms latency is obviously the time between pressing a key and the emission of sound from a "pipe". But this in itself has no scientific meaning. How is the time of pressing a key measured? How is the time a sound is emitted measured? Both must be defined before latency can be properly measured. I am not aware of any scientific definition of either.

When you press an organ key, at some point in its depression it opens a valve to pass air to a pipe. All this takes time so at what point is the key defined as being pressed?

When an organ pipe begins sounding the sound starts from nothing and gradually builds up to a steady state. At what point is the sound defined as being emitted (for the purpose of latency measurement). This can make an enormous difference to a latency measurement.

I am not expecting this to be resolved here but would be interested in other peoples thoughts. Especially whether they believe latency figures quoted for organs have much meaning.

csw900

**Comment by newVideophile** [comment]
A while ago, I was working on a fast piece at full speed, and I noticed that I was making a lot more mistakes than usual. After wondering why I was suddenly so bad, I had the presence of mind to look at the latency in GO. The value had risen to 40 ms, because (for some reason) the settings had changed. I reset my original configuration, with latency displayed at 20 ms, and started playing correctly again!

So, whatever GO's way of measuring latency, the value it displays makes sense.

Personally, I see 4 steps:
1) The organist decides to press the key
2) From a certain position, the key "understands" that it is depressed
3) The command then travels (mechanically/electrically/electronically, etc.) inside the organ until it triggers the opening of a valve
4) The sound unfolds, with a different build-up, of course, depending on the note and stop used, until it becomes perceptible.

For me, total latency is therefore the time between the start of stage 1 and the end of stage 4. 

The latency of step 1 takes place inside the organist and can therefore be removed from the equation.
 
For step 2, the question is essentially to know at what level of depression the key will detect pressure. I suppose this varies from organ to organ. On mine, I've made a 60/40 setting: start playing if key >= 60%, stop if <=40%.

Step 3 poses no problem of definition. In fact, it's the one that best defines the organ's intrinsic performance.

Step 4 is the most problematic, because as you mention, the criterion chosen (i.e. the sound level threshold) can have a considerable influence on the result. Besides, in real life, the perceived latency of a note will not be the same if played alone or with other notes at the same time.

My two cents!

**Reply by willeke1234** [reply]
Is there a step 5? The sound goes from the pipe to the organist.

**Reply by midimusic** [reply]
Yes the time the sound takes to travel from the pipe to the organist can be significant in a large room such as a church.
csw900

**Comment by midimusic** [comment]
Well it looks as if either nobody else is reading this or they have nothing to say about latency.

newVideophile has expanded my original comment which was intended to be completely general and scientific - hopefully moving towards a proper scientific measurement of latency.

First the mention of figures for latency - apparantly GO "displays" latency - where is this display and is it just a guess or a real measurement? I have not found it. Do you really believe the figures 20ms and 40ms? I have doubts..

If we are now talking only about GO, "Step 3" is the only area where GO could attempt to measure its own latency. It would need to start a high resolution timer when it received a "note on" midi message from rtmidi and stop it when it told the operating system's sound driver to play the note. It would then need to display the result. This would all be possible but as far as I know it does not do this.

I remember the days before computers had hardware sound systems. Latency could easily be 500ms - making playing music from a keyboard next to impossible. Nearly all of the improvement since then has been due to the improvements in computer hardware not in VPOs (software organs).

csw900

**Reply by willeke1234** [reply]
According to the help "Audio/Sound Output State tries to displays actual latency.". Choose Audio/Sound menu -> Sound Output State. It's a calculation with samples per buffer and sample rate.

**Reply by midimusic** [reply]
I found this and it says 69ms on my windows system. It would be useful to know the exact formula GO uses to make this calculation in order to check its believability. On this windows system GO plays plays from a keyboard without ANY NOTICEABLE latency.

csw900

**Comment by oleg68** [comment]
@midimusic 

A way of determining an actual latency might by a special device with a midi-out interface and a microphone that mesures the time interval between sending a key-on midi event and capturing the sound from the speaker. I don't know about existance of such devices.

We may build a software solution: a program that emulates midi messages and listen for an audio input. It eliminates a portion of latencies for midi encoding, converting midi to usb, receiving a usb and decoding midi, but it is usually quite small (unless we use usb over wifi/ip tunelling).

This testing program may listen not an audio input from a real microphone, but a virtual audio input connected with a GrandOrgue audio output. In this case we don't take into account a latency added by an audio device. Usually professional or semi professional audio cards add very small latency, but audiophile DACs or HDMI sound output with an external AV-receiver may add quite large latency that will not be measured in this case.

Usually the largest part of latency is brought by GrandOrgue and other sound processing software behind it, because they need to fill an audio buffer and pass it to the next step of audio peocessing chain. So increasing a sample rate, decreasing a buffer size and eliminating extra steps of the audio processing chain decrease the latency.

20ms and 40ms are not precise and they do not reflect the actual latency. 

Usually a high latency does not bring any problems for playing midi files, but it becomes an obstacle for a live organ performation, that @newVideophile mentioned above.

Real pipe organs also bring some latency for playing. Especially it is true for some pneumatic-tracker ogans, ex. built by Walker.

Usually low notes of pedal divisions have long attacks: they require a significant time to grow the sound apmplitude after they started playing. You can see this by opening their attack wav-samples with a sound editor, ex. `Audacity`.

**Reply by midimusic** [reply]
Many years ago I tried this method of properly measuring latency. All you need is a microphone and sound recorder. Hit the key with something that makes a noise. The resulting recording is then displayed in an audio analyser and the times can be easily determined. However the rise time of the audio waveform is by far the longest time measured and there is no standard to say at what point the audio began. At the first instance the wave can be seen it will be too soft to hear - so at what point does the sound become audible? I would suggest that 70% of the final steady state would be reasonable.

This is a complex subject and as I said above I do not know of any standards.

csw900

**Reply by midimusic** [reply]
I have just done a brief test using GO demo and the results are:
Time to first visual evidence of the note (middle A) around 60 ms.
Time for note to become audible around 100ms.
Buffer size = 128 with 44100 sps stereo.

csw900

**Comment by midimusic** [comment]
Regarding the GO sound output buffer size. My previous experiences with this indicated that the optimum buffer size for all systems is 128. I am disappointed to note that the old default of 1024 (which does not work properly on any system; appears to be still in place and continues to fool others as well as me). This high value (1024) does compromise the maximum speed at which the organ can be played - including the playback of midi files.

I have just confirmed on my windows system the size of this buffer has no significant effect on latency - I can find no noticeable change in human detectable latency for any values betwee 128 and 1024. This is possibly confirmed by "Sound Output State" readings which for my system are as follows for sample rate 44100  (Note that my careful spacing of this table has been ruined by Github's ignoring white space characters!!).
Buffer size,   Latency ms
4096,            69
2048,            69
1024,            69
 512,             58
 256,             58
 128,             52
  64,              50
  32,              50
  16,              50

As I said above it would be very useful to find out from the code how these values are calculated. At present I have doubts about the validity of all latency figures displayed by GO. Hence the desirability of a proper scientific measurement.

csw900

**Reply by willeke1234** [reply]
The code of the calculation, start with the latency according to the audio device:

	if (latency < m_SamplesPerBuffer / (double)m_SampleRate)
		latency = m_SamplesPerBuffer / (double)m_SampleRate;
	if (latency < 2 * m_SamplesPerBuffer / (double)m_SampleRate)
		latency += m_SamplesPerBuffer / (double)m_SampleRate;

I get
Buffer size, Latency ms
1024, 46
512, 23
256, 11
128, 5
64, 3
32, 1
16, 1

**Reply by midimusic** [reply]
@willeke1234 

Thanks for your input, however I think on the third line your "<" should have been ">". The code  would then makes sense.

Your figures appear to assume that "latency" = 0 which is never going to be true.

I assume that the formula's "latency" is actually the figure for "requested latency" which for me always defaults to 50ms - is this correct? Can I alter this default?

The latency calculation takes no account of the processing time from midi in to creation of audio - also it takes no account of the speed of the computer.

Also it is compl;etely false to assume that doubling the sample rate will halve the latency because the processing time will then be doubled and the two will cancel - (to a first approximation).

csw900

**Reply by willeke1234** [reply]
The code is [here](https://github.com/GrandOrgue/grandorgue/blob/2c42bb11cc5948fb21aa927f7f2243fd86d88fd5/src/grandorgue/sound/ports/GOSoundPort.cpp#L44C61-L44C61). The "Sound Output State" latency starts with the actual audio stream latency (maybe a wrong value). The minimum latency is one buffer. If the latency < 2 buffers then one buffer is added.

The latency of my audio device according to GO is 1.6, truncated to 1. The Apple documentation says "the AudioDevice's AudioStreams may have additional latency so they should be queried as well.". I don't see this in the RtAudio code so the number will be too low.

De "requested latency" default of 50 is hard coded.

From the Help:
>**Audio driver selection**
GrandOrgue feature two audio backends:
_RtAudio_ 
selects the number of buffers based on the desired latency setting. The "samples per buffer" and sample rate setting must be supported by the soundcard/driver, otherwise sound rendering will break.
_PortAudio_ 
uses the desired latency setting, the "samples per buffer" setting and the capabilities of the soundcard/driver to determine the operational buffering it will use. Such drivers are marked with (PA) in its name.

>_Audio/Sound Output State_ tries to displays actual latency. Note that the estimate can be based on numbers provided by the hardware and the drivers and may vary. RtAudio is more likely to display too low numbers.

RtAudio says:
>The numberOfBuffers parameter can be used to control stream latency in the Windows DirectSound, Linux OSS, and Linux Alsa APIs only. A value of two is usually the smallest allowed. Larger numbers can potentially result in more robust stream performance,  though likely at the cost of stream latency. The value set by the user is replaced during execution of the RtAudio::openStream() function by the value actually used by the system.

So the "requested latency" might be used for the audio stream settings.

Doubling the sample rate will halve the latency because the buffer will be half the time. I think a note has to wait for the next buffer.

**Reply by oleg68** [reply]
@midimusic 

I think that the reason why you have bigger latency than calculated is that you are on Windows and don't use ASIO. I think using Linux/alsa or Windows/asio can decrease the values. Windows audio subsyatem adds some latency.

But if you don't play GrandOrgue with keyboards and use it only for playing midi files, then your latencies are OK.

**Reply by midimusic** [reply]
Thank you both! However I did say this is a complex subject and I do not think at present that any of us really believe the calculated GO figures, least of all me.

@willeke1234 

The results you get from the formula critically depend on the value of "latency" - if it is not the "requested latency" then what is it and where in the code does it come from? I still think that the "<" sign is wrong because the code does not seem to make sense for larger values of "latency".

A higher sample rate will automatically have a longrer latency because the amount of overall sample processing is doubled. Also to get the same results you would have to double the buffer size. Plus  - it is completely pointless unless the original samples were made at the higher rate.

I do not believe that your calculated latency figures have any meaning in reality. Why don't you actually MEASURE your real life latency using a microphone and sound recorder as I have done above for one of my systems. I would like to know the result and I think you will be surprised.

@oleg68 

I do not believe that ASIO has any value on a modern (fast) computer system. There is fundamentally very little difference between windows and linux. The speed of response is mainly down to the hardware sound system and the processor in the computer. However I will measure the real life latency in one of my linux systems and let you know the results.

I do not only "only play midi files". I am perfectly capable of playing from a keyboard and the real life latency I get on windows is perfectly adequate to provide a good playing experience.

Later:
Latency measurement result using Ubuntu -
First appearance of audio around 100ms, GO demo Gamba 8 middle A
Sound audible around 130ms,  at (approx 0.7 of steady state level)
Sample rate 44100
Samples/buffer 128
ALSA (PA) default: 43ms

So Linux is actually worse than Windows for a real measurement. Both were Intel i5 machines.

Even later:
I have also measured Organteq and eplayOrgan, both on Windows and GO is actually the best!!! I suspect HW will be similar to GO because of their similar origins but have not yet tested it.

The next day:
See the following reference for why Windows is probably fastert than Linux for audio:
https://learn.microsoft.com/en-us/windows-hardware/drivers/audio/low-latency-audio

It would be very useful if other readers would MEASURE their actual real life GO latencies and report here.

csw900

**Reply by willeke1234** [reply]
@midimusic

>if it is not the "requested latency" then what is it and where in the code does it come from?

The latency comes from the audio port. Each audio port and OS has its own implementation.

>A higher sample rate will automatically have a longrer latency because the amount of overall sample processing is doubled.

No, if the sample rate (samples/second) is doubled then
- the number of samples in one buffer stays the same
- the processing speed is doubled
- the time it takes to process one buffer is halved
- the latency is lower

>It would be very useful if other readers would MEASURE their actual real life GO latencies and report here.

What exactly is "GO latency" or how much of the total latency is in GO? To my understanding, the Sound Output State latency is the audio processing latency and can be used to see how audio settings affect latency. The other latency doesn't change. The measured latency should be the Sound Output State number plus a fixed number. Measurements (not very precise):
```
Buffer size  GO  Measurement  Difference
       1024  46          100          54
        512  23           68          45
        256  11           61          50
        128   5           52          47
         64   3           47          44
         32   1           45          44
         16   1           49          48
```

**Reply by midimusic** [reply]
@willeke1234 
Thanks for you comments. I will MEASURE the effecte of doubling sample  rate in GO using windows and let tou know. As present I think your argument is wrong. As I said above the buffer size must be doubled if you double the sample rate.

To have any value the latency figures quoted MUST consider the total time from pressing a key to hearing the sound.  Is that what YOUR "measured" latency figures are? It looks to me as if you ignored the rise time of the sound wave.

I have now measured HW (I no longer use HW so it is the old free version I measured) and Miditzer 216 on windows using "out of box" default settings. Both were disappointingly well over 100ms. 

Having now measured all of my readily available organs it is clear that any real life latency up to around 125ms will be un-detectable by the organist. I am still in progreaa to find the latency where it is first noticeable to the organist and thus affects the playability of the organ.

I am thinking about writing all this up on my website as I can find no information of this type on line at present.

csw900

**Reply by midimusic** [reply]
@willeke1234 

I have now tried changing the sample rate from 44100 to 96000 using GO demo organ keeping the same 128 buffer size. As far as I can tell it has made no measurable difference at all to latency. (There is quite a lot of noise left over from the on purpose keypress noise so it is difficult to make a really precise measurement).

Surely this should not be surprising because, as I understand it, a .wav file sample has the sample rate it should be played built in to the file's header. If you play it at a different rate it will alter the tuning because that is how tuning works. I strongly suspect that GO's audio sample rate setting does absolutely nothing. Can you find it in the code and check out what it actually does?

csw900

**Reply by willeke1234** [reply]
@midimusic 

>It looks to me as if you ignored the rise time of the sound wave.

Yes I did, and I measured from a mouse click on the panel manual. I tried to prove my theory. The rise time of the sound wave is always the same.

>Having now measured all of my readily available organs it is clear that any real life latency up to around 125ms will be un-detectable by the organist. I am still in progreaa to find the latency where it is first noticeable to the organist and thus affects the playability of the organ.

Organists get used to latency. When I (hobby organist) first tried GO, I couldn't play because I wasn't used to latency. Now I don't notice.

>I can find no information of this type on line

Maybe not what you're looking for but Google found
[How does a pipe organist deal with latency or delay?](https://music.stackexchange.com/questions/30454)
[Which Buffer Size Setting Should I Use in My DAW?](https://www.sweetwater.com/sweetcare/articles/which-buffer-size-setting-should-i-use-in-my-daw)

**Reply by midimusic** [reply]
@willeke1234 

Measuring from a mouse click on the screen does not make a proper measurement - organists do not play by using a mouse to click on keys.

The rise time of the sound wave depends upon the rise time of the sample you are playing and is most definately NOT always the same. In order to make a properly repeatable latency measurement you must indicate what note you played and which stop you used as well as which organ you are testing. I always use middle A with a flute like stop as this gives a clear repeatable waveform.

I now know that real measured latency has to be something around or above 150ms before it can be detected by playing keys on the organ. For all of the organs I have measured the real measured latency clusters around 100ms and judging by lack of loud complaints by organists this must be acceptable to them.

Your links do not add much that is not obvious - I was searching for real measurements rather than somebody's opinions. The latency of a real organ in a large room can be really significant, much larger than the measured figures I have made for virtual organs in a very small room.

csw900

**Reply by willeke1234** [reply]
@midimusic
What is your purpose of measuring latency?

**Reply by midimusic** [reply]
@willeke1234 

It is an interesting subject which nobody seems to have done before. There can be no meaningful discussion about latency if nobody has measured it. We now know that the very low figures often quoted have no useful meaning in reality.

It is a pity that a few more GO users have not joined in this discussion. Maybe it is because they are quite happy with 100ms latency.

csw900

**Reply by willeke1234** [reply]
@midimusic 
I will do some measurements of the total latency if you tell me how. The recording looks like this:

![Scherm­afbeelding 2023-09-24 om 15 22 53](https://github.com/GrandOrgue/grandorgue/assets/97431382/f412135c-981d-4bc4-b63a-780b865de9b3)

How do I measure latency? I can see where to start, but how do I determine the end?

**Reply by midimusic** [reply]
Your waveform looks very much the same as those I get. You can see the peaked wave at the start - time begins at the first point you can definately see where this wave began. In this case it is at start of the low level wave before the signal begins to rise steeply. This is the time you hit the key and is time 0.

The silent latency period ends where you first see see the organ note beginning to rise. Measure this time. In this case I would say it is at the start of the downward sloping wave just before the silent crossover point where the organ wave starts rising in amplitude. Keep a record of this time.

The latency time is the total time to the the point where the amplitude of the organ note reaches 70% of its final steady state value. Keep a note of this time.

Save the file with a name that records accurately what it is. Also keep written notes if necessary. You should be able to repeat this measurement at a later date and get the same results.

I hope this is helpful.

csw900

**Reply by willeke1234** [reply]
I don't think I can measure latency reliably. The steady state value is not stable, probably because I'm recording with an iPad. And if it was stable, it is a challenge to determine the 70% point.

**Reply by midimusic** [reply]
@willeke1234 

Many organ pipes do not produce a stable steady state value. However you do ideally need something better than an iPad to make accurate measurements. I used an old iMac with windows 11 and Adobe Audition to make my masurements and still I would not say that the 70% level is easy to determine but at least an approximate value can be found.

When there are no measurements available the best you can do is always better than nothing. A scientist always puts a +-nn% tolerance on his measurements to indicate how accurate they are.

I think you have given up too easily - please try again!! Alternatively if you attach a few waveform files me or someone else can analyse them for you.

csw900

**Comment by willeke1234** [comment]
I think the input level is adjusted automatically, so the levels are unreliable. With a buffer size of 128, my latency is 90-100 ms. With a buffer size of 1024 it is 140-160 ms.

**Comment by manu13320** [comment]
I think latency is difficult to evaluate by ear with organ sounds; a more nervous instrument is more suitable, such as the harpsichord, for example http://duphly.free.fr/en/blanchet.html
An acceptable latency would be 50 ms, 100 ms already seems too much to me.

**Reply by midimusic** [reply]
@manu13320 
I use the blanchet harpsichord sampleset. It sounds OK to me.
I invite you to MEASURE its real latency rather than quoting almost random figures. A harpsichord's shorter rise time will automatically make the measured latency lower. But it will still depend mainly upon the organ it is being played on.
csw900
